cdc/scheduler/schedule_dispatcher.go

```diff
@@ -0,0 +1,624 @@
+}
+
+const (
+	// captureCountUninitialized is a placeholder for an unknown total capture count.
+	captureCountUninitialized = -1
+)
+
+// BaseScheduleDispatcher implements the basic logic of a ScheduleDispatcher.
+// For it to be directly useful to the Owner, the Owner should implement it own
+// ScheduleDispatcherCommunicator.
+type BaseScheduleDispatcher struct {
+	mu            sync.Mutex
+	tables        *util.TableSet                         // information of all actually running tables
+	captures      map[model.CaptureID]*model.CaptureInfo // basic information of all captures
+	captureStatus map[model.CaptureID]*captureStatus     // more information on the captures
+	checkpointTs  model.Ts                               // current checkpoint-ts
+
+	moveTableManager moveTableManager
+	balancer         balancer
+
+	lastTickCaptureCount int
+	needRebalance        bool
+
+	// read only fields
+	changeFeedID model.ChangeFeedID
+	communicator ScheduleDispatcherCommunicator
+	logger       *zap.Logger
+}
+
+// NewBaseScheduleDispatcher creates a new BaseScheduleDispatcher.
+func NewBaseScheduleDispatcher(
+	changeFeedID model.ChangeFeedID,
+	communicator ScheduleDispatcherCommunicator,
+	checkpointTs model.Ts,
+) *BaseScheduleDispatcher {
+	// logger is just the global logger with the `changefeed-id` field attached.
+	logger := log.L().With(zap.String("changefeed-id", changeFeedID))
+
+	return &BaseScheduleDispatcher{
+		tables:               util.NewTableSet(),
+		captureStatus:        map[model.CaptureID]*captureStatus{},
+		moveTableManager:     newMoveTableManager(),
+		balancer:             newTableNumberRebalancer(logger),
+		changeFeedID:         changeFeedID,
+		logger:               logger,
+		communicator:         communicator,
+		checkpointTs:         checkpointTs,
+		lastTickCaptureCount: captureCountUninitialized,
+	}
+}
+
+type captureStatus struct {
+	// SyncStatus indicates what we know about the capture's internal state.
+	// We need to know this before we can make decision whether to
+	// dispatch a table.
+	SyncStatus captureSyncStatus
+
+	// Watermark fields
+	CheckpointTs model.Ts
+	ResolvedTs   model.Ts
+	// We can add more fields here in the future, such as: RedoLogTs.
+}
+
+type captureSyncStatus int32
+
+const (
+	// captureUninitialized indicates that we have not sent an `Announce` to the capture yet,
+	// nor has it sent us a `Sync` message.
+	captureUninitialized = captureSyncStatus(iota) + 1
+	// captureSyncSent indicates that we have sent a `Sync` message to the capture, but have had
+	// no response yet.
+	captureSyncSent
+	// captureSyncFinished indicates that the capture has been fully initialized and is ready to
+	// accept `DispatchTable` messages.
+	captureSyncFinished
+)
+
+// Tick implements the interface ScheduleDispatcher.
+func (s *BaseScheduleDispatcher) Tick(
+	ctx context.Context,
+	checkpointTs model.Ts,
+	// currentTables are tables that SHOULD be running given the current checkpoint-ts.
```

Comment:
In this line, 'SHOUD be running given the current checkpoint-ts' is a little bit hard to understand